// @privy-io/react-auth/dist/cjs/CopyableText-DbFeIZI0.js
"use strict";var e=require("react/jsx-runtime"),r=require("lucide-react"),t=require("react"),i=require("styled-components");let o=i.styled.button`
  display: flex;
  align-items: center;
  justify-content: end;
  gap: 0.5rem;

  svg {
    width: 0.875rem;
    height: 0.875rem;
  }
`,n=i.styled.span`
  display: flex;
  align-items: center;
  gap: 0.25rem;
  font-size: 0.875rem;
  color: var(--privy-color-foreground-2);
`,l=i.styled(r.Check)`
  color: var(--privy-color-icon-success);
  flex-shrink: 0;
`,s=i.styled(r.Copy)`
  color: var(--privy-color-icon-muted);
  flex-shrink: 0;
`;exports.CopyButton=({value:r,includeChildren:i,children:c,...a})=>{let[d,u]=t.useState(!1),x=()=>{navigator.clipboard.writeText(r).catch(console.error),u(!0),setTimeout((()=>u(!1)),1500)};/*#__PURE__*/return e.jsxs(e.Fragment,{children:[i?/*#__PURE__*/e.jsx(o,{...a,onClick:x,children:c}):/*#__PURE__*/e.jsx(e.Fragment,{children:c}),/*#__PURE__*/e.jsx(o,{...a,onClick:x,children:d?/*#__PURE__*/e.jsx(n,{children:/*#__PURE__*/e.jsx(l,{})}):/*#__PURE__*/e.jsx(s,{})})]})},exports.CopyableText=function({children:r,iconOnly:i,value:c,hideCopyIcon:a,...d}){let[u,x]=t.useState(!1);/*#__PURE__*/return e.jsxs(o,{...d,onClick:()=>{navigator.clipboard.writeText(c||("string"==typeof r?r:"")).catch(console.error),x(!0),setTimeout((()=>x(!1)),1500)},children:[r," ",u?/*#__PURE__*/e.jsxs(n,{children:[/*#__PURE__*/e.jsx(l,{})," ",!i&&"Copied"]}):!a&&/*#__PURE__*/e.jsx(s,{})]})};
