// @privy-io/react-auth/dist/cjs/LinkPasskeyScreen-o3ApkBow.js
"use strict";var e=require("react/jsx-runtime"),r=require("lucide-react"),i=require("react"),t=require("styled-components"),n=require("./useActiveWallet-GeBabaGt.js"),o=require("./TodoList-Cii_KrAk.js"),s=require("./internal-context-BJv4f_bO.js"),a=require("./privy-context-D5vU7B4j.js"),c=require("./usePrivy-5hOFCOCA.js"),l=require("./ScreenLayout-BG4zeMOR.js");require("zustand"),require("./context-COBUz8AV.js"),require("@privy-io/js-sdk-core"),require("tinycolor2"),require("viem"),require("viem/utils"),require("./use-unlink-wallet-B4w30Xwo.js"),require("./get-entropy-details-for-user-CaCZGwaE.js"),require("eventemitter3"),require("react-device-detect"),require("./events-context-Di6--rDg.js"),require("ofetch"),require("@privy-io/are-addresses-equal"),require("@privy-io/popup"),require("./paths-DizMb-lU.js"),require("jose"),require("./ModalHeader-CW-SKCyT.js"),require("@heroicons/react/24/outline/ArrowLeftIcon"),require("@heroicons/react/24/outline/ArrowRightIcon"),require("@heroicons/react/24/outline/QuestionMarkCircleIcon"),require("@heroicons/react/24/outline/XMarkIcon"),require("./Screen-BkDy7Wxy.js"),require("./index-ByFhjjzH.js");const d=({passkeys:i,isLoading:t,errorReason:n,success:o,expanded:s,onLinkPasskey:a,onUnlinkPasskey:c,onExpand:d,onBack:h,onClose:x})=>o?/*#__PURE__*/e.jsx(l.ScreenLayout,{title:"Passkeys updated",icon:r.CheckCircle,iconVariant:"success",primaryCta:{label:"Done",onClick:x},onClose:x,watermark:!0}):s?/*#__PURE__*/e.jsx(l.ScreenLayout,{icon:r.FingerprintIcon,title:"Your passkeys",onBack:h,onClose:x,watermark:!0,children:/*#__PURE__*/e.jsx(p,{passkeys:i,expanded:s,onUnlink:c,onExpand:d})}):/*#__PURE__*/e.jsx(l.ScreenLayout,{icon:r.FingerprintIcon,title:"Set up passkey verification",subtitle:"Verify with passkey",primaryCta:{label:"Add new passkey",onClick:a,loading:t},onClose:x,watermark:!0,helpText:n||void 0,children:0===i.length?/*#__PURE__*/e.jsx(y,{}):/*#__PURE__*/e.jsx(u,{children:/*#__PURE__*/e.jsx(p,{passkeys:i,expanded:s,onUnlink:c,onExpand:d})})});let u=t.styled.div`
  margin-bottom: 12px;
`,p=({passkeys:t,expanded:o,onUnlink:s,onExpand:a})=>{let[c,l]=i.useState([]),d=o?t.length:2;/*#__PURE__*/return e.jsxs("div",{children:[/*#__PURE__*/e.jsx(g,{children:"Your passkeys"}),/*#__PURE__*/e.jsxs(m,{children:[t.slice(0,d).map((i=>{/*#__PURE__*/return e.jsxs(q,{children:[/*#__PURE__*/e.jsxs("div",{children:[/*#__PURE__*/e.jsx(f,{children:(t=i,t.authenticatorName?t.createdWithBrowser?`${t.authenticatorName} on ${t.createdWithBrowser}`:t.authenticatorName:t.createdWithBrowser?t.createdWithOs?`${t.createdWithBrowser} on ${t.createdWithOs}`:`${t.createdWithBrowser}`:"Unknown device")}),/*#__PURE__*/e.jsxs(j,{children:["Last used:"," ",(i.latestVerifiedAt??i.firstVerifiedAt)?.toLocaleString()??"N/A"]})]}),/*#__PURE__*/e.jsx(C,{disabled:c.includes(i.credentialId),onClick:()=>(async e=>{l((r=>r.concat([e]))),await s(e),l((r=>r.filter((r=>r!==e))))})(i.credentialId),children:c.includes(i.credentialId)?/*#__PURE__*/e.jsx(n.ButtonLoader,{}):/*#__PURE__*/e.jsx(r.Trash2,{size:16})})]},i.credentialId);var t})),t.length>2&&!o&&/*#__PURE__*/e.jsx(k,{onClick:a,children:"View all"})]})]})},y=()=>/*#__PURE__*/e.jsxs(o.TodoList,{style:{color:"var(--privy-color-foreground)"},children:[/*#__PURE__*/e.jsx(o.TodoItem,{children:"Verify with Touch ID, Face ID, PIN, or hardware key"}),/*#__PURE__*/e.jsx(o.TodoItem,{children:"Takes seconds to set up and use"}),/*#__PURE__*/e.jsx(o.TodoItem,{children:"Use your passkey to verify transactions and login to your account"})]});const h={component:()=>{let{user:r}=a.usePrivyContext(),{unlink:t}=c.useUnlinkPasskey(),{linkWithPasskey:n,closePrivyModal:o}=s.usePrivyInternal(),l=r?.linkedAccounts.filter((e=>"passkey"===e.type)),[u,p]=i.useState(!1),[y,h]=i.useState(""),[x,v]=i.useState(!1),[k,m]=i.useState(!1);return i.useEffect((()=>{0===l.length&&m(!1)}),[l.length]),/*#__PURE__*/e.jsx(d,{passkeys:l,isLoading:u,errorReason:y,success:x,expanded:k,onLinkPasskey:()=>{p(!0),n().then((()=>v(!0))).catch((e=>{if(e instanceof s.PrivyError){if(e.privyErrorCode===s.PrivyErrorCode.CANNOT_LINK_MORE_OF_TYPE)return void h("Cannot link more passkeys to account.");if(e.privyErrorCode===s.PrivyErrorCode.PASSKEY_NOT_ALLOWED)return void h("Passkey request timed out or rejected by user.")}h("Unknown error occurred.")})).finally((()=>{p(!1)}))},onUnlinkPasskey:async e=>(p(!0),await t({credentialId:e}).then((()=>v(!0))).catch((e=>{e instanceof s.PrivyError&&e.privyErrorCode===s.PrivyErrorCode.MISSING_MFA_CREDENTIALS?h("Cannot unlink a passkey enrolled in MFA"):h("Unknown error occurred.")})).finally((()=>{p(!1)}))),onExpand:()=>m(!0),onBack:()=>m(!1),onClose:()=>o()})}},x=t.styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 180px;
  height: 90px;
  border-radius: 50%;
  svg + svg {
    margin-left: 12px;
  }
  > svg {
    z-index: 2;
    color: var(--privy-color-accent) !important;
    stroke: var(--privy-color-accent) !important;
    fill: var(--privy-color-accent) !important;
  }
`;let v=t.css`
  && {
    width: 100%;
    font-size: 0.875rem;
    line-height: 1rem;

    /* Tablet and Up */
    @media (min-width: 440px) {
      font-size: 14px;
    }

    display: flex;
    gap: 12px;
    justify-content: center;

    padding: 6px 8px;
    background-color: var(--privy-color-background);
    transition: background-color 200ms ease;
    color: var(--privy-color-accent) !important;

    :focus {
      outline: none;
      box-shadow: none;
    }
  }
`;const k=t.styled.button`
  ${v}
`;let m=t.styled.div`
  display: flex;
  flex-direction: column;
  align-items: stretch;
  gap: 0.8rem;
  padding: 0.5rem 0rem 0rem;
  flex-grow: 1;
  width: 100%;
`,g=t.styled.div`
  line-height: 20px;
  height: 20px;
  font-size: 1em;
  font-weight: 450;
  display: flex;
  justify-content: flex-beginning;
  width: 100%;
`,f=t.styled.div`
  font-size: 1em;
  line-height: 1.3em;
  font-weight: 500;
  color: var(--privy-color-foreground-2);
  padding: 0.2em 0;
`,j=t.styled.div`
  font-size: 0.875rem;
  line-height: 1rem;
  color: #64668b;
  padding: 0.2em 0;
`,q=t.styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 1em;
  gap: 10px;
  font-size: 0.875rem;
  line-height: 1rem;
  text-align: left;
  border-radius: 8px;
  border: 1px solid #e2e3f0 !important;
  width: 100%;
  height: 5em;
`,w=t.css`
  :focus,
  :hover,
  :active {
    outline: none;
  }
  display: flex;
  width: 2em;
  height: 2em;
  justify-content: center;
  align-items: center;
  svg {
    color: var(--privy-color-error);
  }
  svg:hover {
    color: var(--privy-color-foreground-3);
  }
`,C=t.styled.button`
  ${w}
`;exports.DoubleIconWrapper=x,exports.LinkButton=k,exports.LinkPasskeyScreen=h,exports.LinkPasskeyView=d,exports.default=h;
