// @privy-io/react-auth/dist/cjs/SelectSourceAsset-CK1O6oYR.js
"use strict";var e=require("./use-unlink-wallet-B4w30Xwo.js"),r=require("react/jsx-runtime"),t=require("lucide-react"),o=require("react"),l=require("styled-components"),n=require("./index-B_FKAS5Z.js"),i=require("./ScreenLayout-BG4zeMOR.js");let s=l.styled.span`
  background-color: var(--privy-color-background);
  width: 100%;
  text-align: center;
  border: none;
  font-kerning: none;
  font-feature-settings: 'calt' off;
  display: flex;
  justify-content: center;
  align-items: flex-start;
  cursor: pointer;

  &:focus {
    outline: none !important;
    border: none !important;
    box-shadow: none !important;
  }

  && {
    color: var(--privy-color-foreground);
    font-size: 3.75rem;
    font-style: normal;
    font-weight: 600;
    line-height: 5.375rem;
  }
`,a=l.styled.span`
  color: var(--privy-color-foreground);
  font-kerning: none;
  font-feature-settings: 'calt' off;
  font-size: 1rem;
  font-style: normal;
  font-weight: 600;
  line-height: 1.5rem;
  margin-top: 0.75rem;
`,d=l.styled.button`
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: auto;
  gap: 0.5rem;
  border: 1px solid var(--privy-color-border-default);
  border-radius: var(--privy-border-radius-full);

  && {
    margin: auto;
    padding: 0.5rem 1rem;
  }
`,c=l.styled.div`
  svg {
    width: 1rem;
    height: 1rem;
    border-radius: var(--privy-border-radius-full);
    overflow: hidden;
    border: solid 0.1px var(--privy-color-border-default);
  }
`,u=l.styled.span`
  color: var(--privy-color-foreground);
  font-kerning: none;
  font-feature-settings: 'calt' off;
  font-size: 0.875rem;
  font-style: normal;
  font-weight: 500;
  line-height: 1.375rem;
`,p=l.styled.div`
  color: var(--privy-color-foreground);

  svg {
    width: 1.25rem;
    height: 1.25rem;
  }
`;let f=l.styled.div`
  display: flex;
  flex-direction: column;
  gap: 0.75rem;
  width: 100%;
  max-height: 20.875rem;
  overflow-y: auto;
  scrollbar-width: none;

  &::-webkit-scrollbar {
    display: none;
  }
`,h=l.styled.button`
  border-color: var(--privy-color-border-default);
  border-width: 1px;
  border-radius: var(--privy-border-radius-mdlg);
  border-style: solid;
  display: flex;

  && {
    padding: 0.75rem 1rem;
  }
`,y=l.styled.div`
  display: flex;
  align-items: center;
  gap: 1rem;
  width: 100%;
`,m=l.styled.div`
  svg {
    width: 2.25rem;
    height: 2.25rem;
    border-radius: var(--privy-border-radius-full);
    overflow: hidden;
    border: solid 0.1px var(--privy-color-border-default);
  }
`,g=l.styled.div`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  gap: 0.125rem;
`,v=l.styled.span`
  color: var(--privy-color-foreground);
  font-size: 0.875rem;
  font-weight: 600;
  line-height: 1.25rem;
`,x=l.styled.span`
  color: var(--privy-color-foreground-3);
  font-size: 0.75rem;
  font-weight: 400;
  line-height: 1.125rem;
`;exports.AmountInput=({currency:e="usd",value:t,onChange:l,inputMode:i="decimal",autoFocus:d})=>{let[c,u]=o.useState("0"),p=o.useRef(null),f=t??c,h=n.SUPPORTED_CURRENCIES[e]?.symbol??"$",y=o.useCallback((e=>{let r=e.target.value,t=(r=r.replace(/[^\d.]/g,"")).split(".");t.length>2&&(r=t[0]+"."+t.slice(1).join("")),2===t.length&&t[1].length>2&&(r=`${t[0]}.${t[1].slice(0,2)}`),r.length>1&&"0"===r[0]&&"."!==r[1]&&(r=r.slice(1)),(""===r||"."===r)&&(r="0"),l?l(r):u(r)}),[l]),m=o.useCallback((e=>{!(["Delete","Backspace","Tab","Escape","Enter",".","ArrowLeft","ArrowRight","ArrowUp","ArrowDown","Home","End"].includes(e.key)||(e.ctrlKey||e.metaKey)&&["a","c","v","x"].includes(e.key.toLowerCase()))&&(e.key>="0"&&e.key<="9"||e.preventDefault())}),[]),g=o.useMemo((()=>(f.includes("."),f)),[f]);/*#__PURE__*/return r.jsxs(s,{onClick:()=>p.current?.focus(),children:[/*#__PURE__*/r.jsx(a,{children:h}),g,/*#__PURE__*/r.jsx("input",{ref:p,type:"text",inputMode:i,value:g,onChange:y,onKeyDown:m,autoFocus:d,placeholder:"0",style:{width:1,height:"1rem",opacity:0,alignSelf:"center",fontSize:"1rem"}}),/*#__PURE__*/r.jsx(a,{style:{opacity:0},children:h})]})},exports.CurrencySelector=({selectedAsset:e,onEditSourceAsset:o})=>{let{icon:l}=n.SUPPORTED_CURRENCIES[e];/*#__PURE__*/return r.jsxs(d,{onClick:o,children:[/*#__PURE__*/r.jsx(c,{children:l}),/*#__PURE__*/r.jsx(u,{children:e.toLocaleUpperCase()}),/*#__PURE__*/r.jsx(p,{children:/*#__PURE__*/r.jsx(t.ChevronDown,{})})]})},exports.SelectSourceAsset=({opts:e,isLoading:t,onSelectSource:o})=>/*#__PURE__*/r.jsx(i.ScreenLayout,{showClose:!1,showBack:!0,onBack:()=>o(e.source.selectedAsset),title:"Select currency",children:/*#__PURE__*/r.jsx(f,{children:e.source.assets.map((e=>{let{icon:l,name:i}=n.SUPPORTED_CURRENCIES[e];/*#__PURE__*/return r.jsx(h,{onClick:()=>o(e),disabled:t,children:/*#__PURE__*/r.jsxs(y,{children:[/*#__PURE__*/r.jsx(m,{children:l}),/*#__PURE__*/r.jsxs(g,{children:[/*#__PURE__*/r.jsx(v,{children:i}),/*#__PURE__*/r.jsx(x,{children:e.toLocaleUpperCase()})]})]})},e)}))})}),exports.poll=async({operation:r,until:t,delay:o,interval:l,attempts:n,signal:i})=>{let s,a,d=0;for(;d<n;){if(i?.aborted)return{status:"aborted",result:s,attempts:d,error:a};d++;try{if(a=void 0,s=await r(),t(s))return{status:"success",result:s,attempts:d};d<n&&await e.sleep(l)}catch(r){r instanceof Error&&(a=r),d<n&&await e.sleep(l)}}return{status:"max_attempts",result:s,attempts:d,error:a}};
