// @privy-io/api-types/resources/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Yield = exports.Webhooks = exports.Wallets = exports.WalletActions = exports.Users = exports.Transactions = exports.Swaps = exports.Shared = exports.Policies = exports.Organizations = exports.KrakenEmbed = exports.KeyQuorums = exports.Intents = exports.Funding = exports.EmbeddedWallets = exports.CrossApp = exports.ClientAuth = exports.Apps = exports.Analytics = exports.Aggregations = exports.Accounts = void 0;
var accounts_1 = require("./accounts.js");
Object.defineProperty(exports, "Accounts", { enumerable: true, get: function () { return accounts_1.Accounts; } });
var aggregations_1 = require("./aggregations.js");
Object.defineProperty(exports, "Aggregations", { enumerable: true, get: function () { return aggregations_1.Aggregations; } });
var analytics_1 = require("./analytics.js");
Object.defineProperty(exports, "Analytics", { enumerable: true, get: function () { return analytics_1.Analytics; } });
var apps_1 = require("./apps.js");
Object.defineProperty(exports, "Apps", { enumerable: true, get: function () { return apps_1.Apps; } });
var client_auth_1 = require("./client-auth.js");
Object.defineProperty(exports, "ClientAuth", { enumerable: true, get: function () { return client_auth_1.ClientAuth; } });
var cross_app_1 = require("./cross-app.js");
Object.defineProperty(exports, "CrossApp", { enumerable: true, get: function () { return cross_app_1.CrossApp; } });
var embedded_wallets_1 = require("./embedded-wallets.js");
Object.defineProperty(exports, "EmbeddedWallets", { enumerable: true, get: function () { return embedded_wallets_1.EmbeddedWallets; } });
var funding_1 = require("./funding.js");
Object.defineProperty(exports, "Funding", { enumerable: true, get: function () { return funding_1.Funding; } });
var intents_1 = require("./intents.js");
Object.defineProperty(exports, "Intents", { enumerable: true, get: function () { return intents_1.Intents; } });
var key_quorums_1 = require("./key-quorums.js");
Object.defineProperty(exports, "KeyQuorums", { enumerable: true, get: function () { return key_quorums_1.KeyQuorums; } });
var kraken_embed_1 = require("./kraken-embed.js");
Object.defineProperty(exports, "KrakenEmbed", { enumerable: true, get: function () { return kraken_embed_1.KrakenEmbed; } });
var organizations_1 = require("./organizations.js");
Object.defineProperty(exports, "Organizations", { enumerable: true, get: function () { return organizations_1.Organizations; } });
var policies_1 = require("./policies.js");
Object.defineProperty(exports, "Policies", { enumerable: true, get: function () { return policies_1.Policies; } });
var shared_1 = require("./shared.js");
Object.defineProperty(exports, "Shared", { enumerable: true, get: function () { return shared_1.Shared; } });
var swaps_1 = require("./swaps.js");
Object.defineProperty(exports, "Swaps", { enumerable: true, get: function () { return swaps_1.Swaps; } });
var transactions_1 = require("./transactions.js");
Object.defineProperty(exports, "Transactions", { enumerable: true, get: function () { return transactions_1.Transactions; } });
var users_1 = require("./users.js");
Object.defineProperty(exports, "Users", { enumerable: true, get: function () { return users_1.Users; } });
var wallet_actions_1 = require("./wallet-actions.js");
Object.defineProperty(exports, "WalletActions", { enumerable: true, get: function () { return wallet_actions_1.WalletActions; } });
var wallets_1 = require("./wallets/wallets.js");
Object.defineProperty(exports, "Wallets", { enumerable: true, get: function () { return wallets_1.Wallets; } });
var webhooks_1 = require("./webhooks.js");
Object.defineProperty(exports, "Webhooks", { enumerable: true, get: function () { return webhooks_1.Webhooks; } });
var yield_1 = require("./yield.js");
Object.defineProperty(exports, "Yield", { enumerable: true, get: function () { return yield_1.Yield; } });
